### PR TITLE
Looking for input on api-docs for client-files

### LINKED ISSUE #364: Client Files Backend
# Requirements

- Downloading the file must be authenticated/authorized. This is because immigration files are sensitive.
- Files are associated with exactly one client
- Unlimited number of files can be uploaded to each client
- Maximum file size limitation is completely okay as long as it is at least 
- Downloading the file should have the correct Content-Disposition so that it triggers a download instead of the browser trying to parse/display it.

# Implementation approach

- Please submit a PR to the api-docs folder of this repo with a new markdown documentation file for your API before working on the code
- I don't mind whether the file store is S3 or MySQL, as long as it works and isn't extremely slow or pricey or anything

### PR DESCRIPTION
**Pull request by seanbw89**
resolves #364

### CODE REVIEW

File: api-docs/client-files.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
+# Client files
+
+This api is provided to upload and keep track of client files.
+
+## Get a single client file
+
+```http
+GET /api/clients/:file/:id/files
```
Comment (joeldenning) [inline api-docs/client-files.md]: comunidades-unidas has RESTful APIs. The term REST is often bastardized to mean a lot of things -- it started out as an academic paper in a journal and has morphed into meaning a lot of different things. A main concept in REST, though, is [resources](https://www.tutorialspoint.com/restful/restful_resources.htm).

The way you structure your url is determined by the resources that are involved in the API. In this case, we have two resources: `clients` and `files`. The more common way to structure urls is make the url readable left-to-right where the id parameter comes after the resource name.

In other words, let's rearrange the url here:

```suggestion
GET /api/clients/:clientId/files/:fileId
```

File: api-docs/client-files.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "lastUpdatedBy": {
+    "firstName": "Sean",
+    "lastName": "White",
+    "fullName": "Sean White",
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "fileName": "file-1",
+  "filePath": "file://host/path"
```
Comment (joeldenning) [inline api-docs/client-files.md]: What is this filePath going to be? Strings that start with `file://` usually refer to files that are local to the computer. For the user of this API, they won't have this file on their hard drive. Is `filePath` the URL to download the file? If we're using signed s3 urls, I might even expect a completely separate api request to get the signed url to download the file (instead of always creating a signed url to download even when you're just viewing the file list).

File: api-docs/client-files.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+```json
+{
+  "createdBy": {
+    "firstName": "Sean",
+    "lastName": "White",
+    "fullName": "Sean White",
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "lastUpdatedBy": {
```
Comment (joeldenning) [inline api-docs/client-files.md]: The files we store won't really get updated. I don't think this `lastUpdatedBy` section should be here.

File: api-docs/client-files.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    "firstName": "Sean",
+    "lastName": "White",
+    "fullName": "Sean White",
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "fileName": "file-1",
```
Comment (joeldenning) [inline api-docs/client-files.md]: Any other metadata we want to add to this API? The file size could also be useful to send so that that can be displayed to the user before they attempt to download.

File: api-docs/client-files.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+{
+  "errors": ["Could not find client with id 2"]
+}
+```
+
+If there is no file with the provided id, you will get a 404 HTTP response, with the following error
+
+```json
+{
+  "errors": ["Could not find file with #123"]
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
  "errors": ["Could not find file with id 123"]
```

File: api-docs/client-files.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+}
+```
+
+&nbsp;
```
Comment (joeldenning) [inline api-docs/client-files.md]: What's the purpose of the non-breaking spaces? Is it to create a new line? Isn't the `---` below good enough to separate the APIs?

File: api-docs/client-files.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+      "lastUpdatedBy": {
+        "firstName": "Sean",
+        "lastName": "White",
+        "fullName": "Sean White",
+        "timestamp": "2019-11-06T06:00:00.000Z"
+      },
+      "fileName": "file-1",
+      "filePath": "file://host/path"
+    },
+    "..."
```
Comment (joeldenning) [inline api-docs/client-files.md]: I prefer that the API examples be copy-pastable when mocking the API. Let's remove this line so that that is true of this example. The `files` property being an array already indicates that it is possible to have more than one object in it.

```suggestion
```

File: api-docs/client-files.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+&nbsp;
+
+---
+
+&nbsp;
+
+## Create a single client file
+
+```http
+POST /api/client/file
```
Comment (joeldenning) [inline api-docs/client-files.md]: RESTful resources (the full url) can and should be reused for listing, creating, and deleting files. We should be using the same url for all of the POST, GET, and DELETE operations.

```suggestion
POST /api/clients/:clientId/files
```

File: api-docs/client-files.md [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+```
+
+### Request
```
Comment (joeldenning) [inline api-docs/client-files.md]: Uploading files via a browser is only possible by using [multipart form data](https://developer.mozilla.org/en-US/docs/Web/API/FormData/Using_FormData_Objects). This means that the request headers for uploading a file must include `Content-Type: multipart/form-data`, and the request body should be the multipart form data format (which is not JSON). We should clarify all of this in the API docs, and also remove the JSON you have here in the request body.

Unless, are you expecting the file upload to go to S3 and this POST request just to be the way the UI tells the web server that they completed the uploading of a file?

File: api-docs/client-files.md [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+
+### Notes
+
+### Validation Error
+
+Validation errors will respond with HTTP status 400.
+
+```json
+{
+  "errors": ["You must provide a id"]
```
Comment (joeldenning) [inline api-docs/client-files.md]: This is not needed, since the clientId should be in the url instead of request body.

File: api-docs/client-files.md [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+```http
+POST /api/client/file
+```
+
+### Request
+
+```json
+{
+  "clientId": 1,
```
Comment (joeldenning) [inline api-docs/client-files.md]: Not needed, since the clientId should be in the request url

File: api-docs/client-files.md [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+      "lastUpdatedBy": {
+        "firstName": "Sean",
+        "lastName": "White",
+        "fullName": "Sean White",
+        "timestamp": "2019-11-06T06:00:00.000Z"
+      },
+      "fileName": "file-1",
+      "filePath": "file://host/path"
+    },
+    "..."
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
```

File: api-docs/client-files.md [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+&nbsp;
+
+---
+
+&nbsp;
+
+## Create multiple client files
+
+```http
+POST /api/client/files
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
POST /api/clients/:clientId/files
```

File: api-docs/client-files.md [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+```
+
+### Request
+
+```json
+{
+  "clientId": "1",
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
```

File: api-docs/client-files.md [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+&nbsp;
+
+---
+
+&nbsp;
+
+## Delete a single client file
+
+```http
+DELETE /api/client/:id/:file
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
DELETE /api/client/:clientId/files/:files
```

File: api-docs/client-files.md [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+}
+```
+
+&nbsp;
+
+---
+
+&nbsp;
+
+## Delete a single client file
```
Comment (joeldenning) [inline api-docs/client-files.md]: Let's add a note to this API clarifying whether this API does a hard delete or soft delete on the file itself.

File: api-docs/client-files.md [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+
+An HTTP 400 status is return if you cannot delete the file
+
+&nbsp;
+
+---
+
+&nbsp;
+
+## Delete all client files
```
Comment (joeldenning) [inline api-docs/client-files.md]: I don't know of a use case for this API. In fact, it is rather dangerous to be able to so easily delete all of the client files.

I also don't think we need a batch delete API (one that deletes by id instead of just all of them) just now for this, either.

File: api-docs/client-files.md [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+```http
+GET /api/clients/:file/:id/files
+```
+
+### Response
+
+#### Success
+
+```json
+{
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: Do we need files to be associated with specific client interactions too? If so can we add `clientInteractionId`

File: api-docs/client-files.md (hunk 6 above)
Comment (seanbw89) [reply api-docs/client-files.md]: The non-breaking spaces, was just for my OCD, I will remove them

File: api-docs/client-files.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (joeldenning) [reply api-docs/client-files.md]: 👍 agreed

File: api-docs/client-files.md [hunk 19]
```diff
@@ -134,18 +130,14 @@ The response will look identical to get a client file
...
   "fileName": "file-1",
-  "filePath": "file://host/path"
+  "fileSize": "10mb",
+  "fileExtension": "jpeg",
+  "fileNote": "some note here"
 }
 ```
 
-### Notes
+**_Notes_**
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: No notes are in this section.

File: api-docs/client-files.md [hunk 20]
```diff
@@ -232,31 +219,15 @@ Validation errors will respond with HTTP status 400.
...
-
-&nbsp;
+**note**
 
-## Delete all client files
-
-```http
-DELETE /api/client/:id/files
-```
+This will ne a hard delete of the file. Should indicate to the client that this is permanant, and cannot be reversed.
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: `be`

File: api-docs/client-files.md [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+
+&nbsp;
+
+## Delete all client files
+
+```http
+DELETE /api/client/:id/files
+```
+
+## Response
```
[the remark this answers is not in the document]
Comment (MeghanBomberger) [reply api-docs/client-files.md]: Looks like we still need this one fixed.

File: api-docs/client-files.md [hunk 22]
```diff
@@ -232,31 +219,15 @@ Validation errors will respond with HTTP status 400.
...
-An HTTP 204 status is returned if the deletion was sucessful
-
-An HTTP 400 status is return if you cannot delete the file
-
-&nbsp;
-
----
-
-&nbsp;
+**note**
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: Different format than the other note headers you have.

File: api-docs/client-files.md [hunk 23]
```diff
@@ -197,21 +190,15 @@ The response will look identical to get all client files execpt that it will onl
...
       },
-      "lastUpdatedBy": {
-        "firstName": "Sean",
-        "lastName": "White",
-        "fullName": "Sean White",
-        "timestamp": "2019-11-06T06:00:00.000Z"
-      },
       "fileName": "file-1",
-      "filePath": "file://host/path"
+      "fileSize": "file://host/path"
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: Should the response data here be the file size info rather than the path?

File: api-docs/client-files.md [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+      },
+      "fileName": "file-1",
+      "fileSize": "file://host/path"
+    },
+    "..."
+  ]
+}
+```
+
+**_Notes_**
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: No notes under this header. If there are no notes for this section it isn't needed.

File: api-docs/client-files.md [hunk 25]
```diff
@@ -0,0 +1,233 @@
...
+
+## Delete a single client file
+
+```http
+DELETE /api/client/:clientId/files/:files
+```
+
+**note**
+
+This will ne a hard delete of the file. Should indicate to the client that this is permanant, and cannot be reversed.
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: `permanent`

File: api-docs/client-files.md [hunk 26]
```diff
@@ -0,0 +1,233 @@
...
+      "fileNote": "Some note here"
+    },
+    "...."
+  ]
+}
+```
+
+### Response
+
+The response will look identical to get all client files execpt that it will only return the array of the current files that you are uploading
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: `except`

File: api-docs/client-files.md [hunk 27]
```diff
@@ -0,0 +1,233 @@
...
+```
+
+### Request
+
+```json
+{
+  "clientId": "1",
+  "files": [
+    {
+      "fileName": "filname.ext",
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: `"filename.ext"`

File: api-docs/client-files.md [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "fileName": "file-1",
+  "fileSize": "10mb",
+  "fileExtension": "jpeg",
+  "fileNote": "some note here"
+}
+```
+
+**_Notes_**
```
Comment (MeghanBomberger) [inline api-docs/client-files.md]: No notes under this header. If there are no notes for this section it isn't needed.

File: api-docs/client-files.md [hunk 29]
```diff
@@ -0,0 +1,318 @@
...
+
+```json
+{
+  "errors": ["Could not find client with id 2"]
+}
+```
+
+---
+
+## Create a client file
```
Comment (joeldenning) [inline api-docs/client-files.md]: We need to describe in this markdown file the full flow for both downloading and uploading a file. It is not clear to the reader of this API documentation that this "Create a client file" api should be called after they've already uploaded the file to S3.

File: api-docs/client-files.md [hunk 30]
```diff
@@ -0,0 +1,318 @@
...
+  "fileName": "fileKey",
+  "fileSize": "10mb",
+  "fileExtension": "pdf",
+  "fileDesc": "invoice"
+}
+```
+
+**NOTES**
+
+The `filename` will be included in the signed url. Not in the uploaded file.
```
Comment (joeldenning) [inline api-docs/client-files.md]: I don't understand this comment

File: api-docs/client-files.md [hunk 31]
```diff
@@ -0,0 +1,262 @@
...
+  "clientId": 1,
+  "fileName": "filename.ext"
+}
+```
+
+### Response
+
+The response will look identical to get a client file
+
+```json
```
[the remark this answers is not in the document]
Comment (joeldenning) [reply api-docs/client-files.md]: 👍 this should have an id

File: api-docs/client-files.md [hunk 32]
```diff
@@ -0,0 +1,318 @@
...
+The `filename` will be included in the signed url. Not in the uploaded file.
+
+### Response
+
+#### Success
+
+```json
+{
+  "clientId": 1,
+  "interactionId": 1,
```
Comment (joeldenning) [inline api-docs/client-files.md]: This api has nothing to do with interactions:

```suggestion
```

File: api-docs/client-files.md [hunk 33]
```diff
@@ -0,0 +1,318 @@
...
+
+```json
+{
+  "errors": ["You must provide a id"]
+}
+```
+
+---
+
+## Create multiple client files
```
Comment (joeldenning) [inline api-docs/client-files.md]: I don't think we need this api at all. We can call the other api multiple times instead of having a batch api

File: api-docs/client-files.md [hunk 34]
```diff
@@ -0,0 +1,318 @@
...
+
+The `filename` will be included in the signed url. Not in the uploaded file.
+
+### Response
+
+#### Success
+
+```json
+{
+  "clientId": 1,
```
Comment (joeldenning) [inline api-docs/client-files.md]: The clientId isn't needed since the code calling this API already specified the clientId in the request -- they already know what the clientId is without the server telling them

File: api-docs/client-files.md [hunk 35]
```diff
@@ -0,0 +1,318 @@
...
+  "createdBy": {
+    "firstName": "Sean",
+    "lastName": "White",
+    "fullName": "Sean White",
+    "timestamp": "2019-11-06T06:00:00.000Z"
+  },
+  "fileName": "fileKey",
+  "fileSize": "10mb",
+  "fileExtension": "pdf",
+  "fileDesc": "invoice"
```
Comment (joeldenning) [inline api-docs/client-files.md]:
```suggestion
```

File: api-docs/client-files.md [hunk 36]
```diff
@@ -0,0 +1,318 @@
...
+  "errors": ["You must provide a id"]
+}
+```
+
+---
+
+## Soft delete a client file
+
+```http
+DELETE /api/clients/:clientId/file/:fileKey
```
Comment (joeldenning) [inline api-docs/client-files.md]: `fileId`, not `fileKey`. We want to allow for two files with the same name, which is only possible if we use the mysql id as the unique thing instead of the fileKey name

**Comment by joeldenning** [comment]
I just resolved all comments that have been addressed or are no longer relevant. Please address all the unresolved comments

File: api-docs/client-files.md [hunk 37]
```diff
@@ -0,0 +1,270 @@
...
+a PUT request to the url like so
+
+```javascript
+axios.put(signedRequest, file, options).then(result => {
+  if (result.data.success) {
+    //Make the request to create a client file /api/clients/:clientId/file
+  }
+});
+```
+
```
Comment (joeldenning) [inline api-docs/client-files.md]: Let's continue explaining here:

> After you upload the file to S3, you need to tell the web server that you did so by calling POST /api/clients/:clientId/files

File: api-docs/client-files.md [hunk 38]
```diff
@@ -0,0 +1,274 @@
...
+After you upload the file to S3, you need to tell the web server that you
+did so by calling `http POST /api/clients/:clientId/files`
+
+For each file that you want to download you would make a request to `/api/clients/:clientId/files/:fileId/signed-downloads` then
+you would make a request on the front end like so `window.location.href = downloadurl`
+
+## Get a signed POST URL
+
+```http
+GET /api/signed-file-uploads
```
Comment (joeldenning) [inline api-docs/client-files.md]: Since you and I talked about how the `fileKey` returned from the server is going to have the client directory and a uuid for filename, we should update this API so that it is specific to a client so that the server knows which directory to put the filekey in.

```suggestion
GET /api/clients/:clientId/signed-file-uploads
```
